Allow remote signer to run without chain backend

guggero opened the pull request:
~Depends on https://github.com/btcsuite/btcwallet/pull/781.~

This fixes/finishes the remote signing in that it makes it no longer required for the signing instance to have its own chain backend. Therefore all address synchronization issues are removed as well, the signing instance can now be completely offline (except for the incoming RPC connection).
Additionally, only the `signrpc` interface needs to be implemented by the remote signer.

Fixes https://github.com/lightningnetwork/lnd/issues/5243

Fixes https://github.com/lightningnetwork/lnd/issues/5321 

Fixes https://github.com/lightninglabs/loop/issues/437

yyforyongyu commented on lnrpc/signrpc/signer_server.go:
```diff
@@ -310,9 +310,10 @@ func (s *Server) SignOutputRaw(ctx context.Context, in *SignReq) (*SignResp,
 		// If a witness script isn't passed, then we can't proceed, as
 		// in the p2wsh case, we can't properly generate the sighash.
+		// For P2WKH we expect the PK script to be sent which the
+		// SignOutputRaw method will use correctly when creating the
```
nit: I think you mean "SignOutputRaw method will be used correctly"?

guggero replied:
No, I mean that SignOutputRaw will handle the PK script correctly, even though it's technically not a PK script. I changed the sentence a bit to make it easier to understand.

yyforyongyu commented on lnwallet/btcwallet/psbt.go:
```diff
@@ -0,0 +1,107 @@
+// finalized successfully.
+//
+// This is a part of the WalletController interface.
+func (b *BtcWallet) FinalizePsbt(packet *psbt.Packet, accountName string) error {
```
Since we are here, unrelated to this PR,

So `btcwallet.FinalizePsbt` is equivalent to bitcoind's `walletprocesspsbt`? That each participant calls `FinalizePsbt` and passes the partially signed tx to the next participant till all sigs are collected?

guggero replied:
Not exactly. The `FinalizePsbt` will add signatures but then will also try to finalize the PSBT which will fail if there are unsigned inputs. The new `SignPsbt` is more like `walletprocesspsbt` in that it tries to sign whatever it can but doesn't fail if the TX isn't complete afterwards (in case more signers are needed).

yyforyongyu commented on lnwallet/btcwallet/signer.go:
```diff
@@ -73,6 +76,94 @@ func deriveFromKeyLoc(scopedMgr *waddrmgr.ScopedKeyManager,
 
+// deriveKeyByBIP32Path derives a key described by a BIP32 path. The path is
+// expected to have 2^31 added to a hardened key in the path.
+func (b *BtcWallet) deriveKeyByBIP32Path(path []uint32) (*btcec.PrivateKey,
```
Curious about the relationship of `btcsuite/btcwallet` and `lnwallet/btcwallet`. Should the HD wallet related implementations live in `btcsuite` since it's quite low level?

guggero replied:
The `btcsuite` one is agnostic of any of our key scopes. So the fact that we have an internal (`1017'`) and two external (`49'` and `84'`) scopes is not known to it. Therefore the logic in the `deriveKeyByBIP32Path()` that switches between the internal and external scopes must be in `lnd`.

yyforyongyu reviewed (commented):
Did a high level pass. Very necessary change as the remote signers can now live in a "cold" environment without running a bitcoin node. Thanks for the effort!

guggero commented:
One question I was pondering: Should we add a new itest mode where every node in the itest gets its own remote signer? Since we don't need a chain backend anymore it should be doable. That would certainly increase the test coverage of the signing. But I'm not sure if that warrants the extra time spent on it (and also the new itest GitHub slot that would be taken up by it).

carlaKC commented on lnwallet/btcwallet/signer.go:
```diff
@@ -73,6 +76,94 @@ func deriveFromKeyLoc(scopedMgr *waddrmgr.ScopedKeyManager,
+	account := path[2] - hdkeychain.HardenedKeyStart
+	change, index := path[3], path[4]
+
+	// Is this an internal key?
```
newbie-nit: the term "internal" makes me think of change addresses in this context (internal vs external)
ignore if we already use this terminology elsewhere, just a thought

guggero replied:
Hmm, good point. I rephrased "internal" and "external" to "custom purpose" and "default BIP" in this context to make it more clear. I hope I caught every occurrence.

carlaKC commented on lnwallet/btcwallet/psbt.go:
```diff
@@ -64,6 +85,164 @@ func (b *BtcWallet) FundPsbt(packet *psbt.Packet, minConfs int32,
+			return input.TweakPrivKey(privKey, u.Value)
+		}
+
+		if bytes.Equal(u.Key, PsbtKeyTypeInputSignatureTweakDouble) {
```
Do we need to fail if `u.Key` isn't `nil / TweakSingle / TweakDouble`?

guggero replied:
No, it's completely optional and we shouldn't fail on any foreign keys. Added a comment with some more context.

carlaKC commented on chainreg/chainregistry.go:
```diff
@@ -608,6 +608,22 @@ func NewPartialChainControl(cfg *Config) (*PartialChainControl, func(), error) {
 			}
 		}
+
+	case "nochainbackend":
```
I haven't thought about this with sufficient detail, but what are the tradeoffs between running all of lnd in `nochainbackend` mode vs trying to pull the signer out into its own binary? Is it a big lift to make the subserver into our primary rpc server?

guggero replied:
Yes, at the moment I think it is quite a big lift (and potentially a big code duplication issue) to extract the signer. I think the main issue would be that we'd need to fully implement the main `lnrpc.Lightning` service (to support calls like `GetInfo` and perhaps `BakeMacaroon`).

Roasbeef replied:
Yeah I think pulling it out into a _distinct_ binary is something we want to do eventually, with the stop gap being thins restricted operating mode. 

Similarly, we should eventually also pull out the watch tower into a distinct binary.

carlaKC commented on docs/remote-signing.md:
````diff
@@ -31,39 +31,13 @@ xxx               xx
-```
+  +-------+--------+              |   +----------------+             |
+          |                       |                                  |
+  +-------v--------+              +----------------------------------+
````
🔥

carlaKC reviewed (commented):
First pass looks good! Will take a more detailed look at tests + `RPCKeyRing` in next look.

carlaKC commented:
> Should we add a new itest mode where every node in the itest gets its own remote signer? Since we don't need a chain backend anymore it should be doable. 

Could be an easier solution to just add a single itest case with 1x remote signer and open channel/ transact close? Will be a bit of additional wiring on the harness for single use but I think that'll give us _good enough_ coverage?

yyforyongyu reviewed (changes requested):
Nice work! I'll need to get more familiar with the hd wallet implemented in `lnwallet` to understand the scope change. Meanwhile, have questions re unit tests and the 256 key families init.

> Should we add a new itest mode where every node in the itest gets its own remote signer?

Then I guess we would need to visit every test to validate the expected behavior of the remote signer?

guggero commented on lnwallet/btcwallet/psbt.go:
```diff
@@ -64,6 +85,164 @@ func (b *BtcWallet) FundPsbt(packet *psbt.Packet, minConfs int32,
+			return fmt.Errorf("error signing input %d: %v", idx,
+				err)
+		}
+		packet.Inputs[idx].PartialSigs = append(
```
As long as only one goroutine accesses the packet at any given time, this should be safe, yes. Or what do you mean by concurrent access?

yyforyongyu replied:
Yeah that's what I meant. Since the packet is just a pointer, was concerning other goroutines might modify it too. Checked the packet and it seems all the data are safe for concurrent access so I think it's ok.

guggero commented on lnwallet/btcwallet/signer.go:
```diff
@@ -73,6 +76,94 @@ func deriveFromKeyLoc(scopedMgr *waddrmgr.ScopedKeyManager,
+		})
+	}
+
+	// Okay, it's a BIP49/84 key, we need to derive it. Interestingly, the
```
Yeah, I guess we should validate that. Added.

guggero commented on lnwallet/btcwallet/psbt_test.go:
```diff
@@ -0,0 +1,319 @@
+	}, {
+		name:      "single key tweaked P2WSH",
+		inputType: singleKeyTweakedP2WSH,
+	}}
```
Nope, that's the last test case. I renamed it to be more clear.

guggero commented on lnwallet/btcwallet/psbt_test.go:
```diff
@@ -0,0 +1,319 @@
+		}}
+		tc.inputType.decorateInput(t, privKey, &packet.Inputs[0])
+
+		// Let the wallet do its job. We expect to be the only signer
```
Good question, but unfortunately the answer is no. Added a comment to the method and a TODO for a refactor (needs another PR in btcwallet that I'd rather not block this PR on at the moment).

guggero commented on lnwallet/rpcwallet/rpcwallet.go:
```diff
@@ -401,48 +219,136 @@ func (r *RPCKeyRing) FinalizePsbt(packet *psbt.Packet, accountName string) error
+// finalized successfully.
+//
+// This is a part of the WalletController interface.
+func (r *RPCKeyRing) FinalizePsbt(packet *psbt.Packet, _ string) error {
```
Good question, but unfortunately the answer is no. Added a comment to the method and a TODO for a refactor (needs another PR in btcwallet that I'd rather not block this PR on at the moment).

guggero commented on lnwallet/btcwallet/btcwallet.go:
```diff
@@ -323,25 +323,21 @@ func (b *BtcWallet) Start() error {
+		// Generate all accounts that we could ever need. This includes
+		// all lnd key families as well as some key families used in
+		// external liquidity tools.
+		for keyFam := uint32(1); keyFam <= 255; keyFam++ {
```
Yes, this will be created for every wallet since there is no way to import a custom key family from the CLI.

yyforyongyu replied:
The how about create the accounts needed only? Not sure about adding this data to every lnd node without the node operator's awareness. And this only adds accounts under 1017/1 right?

guggero replied:
The downside to only adding those we actually need instead of just the first 255 ones is that we keep a list of Loop/Pool accounts/key families hard coded in `lnd`. And if we ever add a new key family to a product we first need to add it to `lnd`. With just using 255 accounts/families we just need to keep any new family below that number.
I think it's a reasonable tradeoff in exchange for an additional ~300kB of data.

> And this only adds accounts under 1017/1 right?

Yes, this only affects the `1017'` key scope.

Roasbeef replied:
Only downside I can think of here is that the initial call to list accounts will list all these other accounts, but that can be addressed primarily with documentation, or if we filter out the ones w/o any active addresses.

guggero commented:
> > Should we add a new itest mode where every node in the itest gets its own remote signer? Since we don't need a chain backend anymore it should be doable.
> 
> Could be an easier solution to just add a single itest case with 1x remote signer and open channel/ transact close? Will be a bit of additional wiring on the harness for single use but I think that'll give us _good enough_ coverage?

Not sure I understand what you mean. Isn't this what's already been done in the `lntest/itest/lnd_remote_signer_test.go` test?

guggero commented on lnwallet/btcwallet/psbt.go:
```diff
@@ -64,6 +85,164 @@ func (b *BtcWallet) FundPsbt(packet *psbt.Packet, minConfs int32,
+	}
+
+	// Go through each input that doesn't have final witness data attached
+	// to it already and try to sign it. We do expect that we're the last
```
Yes, copy/paste artifact.

carlaKC commented:
> Isn't this what's already been done in the lntest/itest/lnd_remote_signer_test.go test?

Yeah ofc, forgot we had that!

carlaKC reviewed (approved):
Reviewed to the best of my ability + tested locally (np2wkh/ p2wkh + channel open). 

I haven't done much wallet work, so can't be confident that I'd pick up any subtle bugs, so you might want to pick up another reviewer with more domain knowledge. 

One thing I did notice while testing is that `watch-only` won't shut down if `signer` is down. Could use `Criticalf` if signing fails, or add a health check if we're running with remote signer.

yyforyongyu commented on lnwallet/btcwallet/psbt.go:
```diff
@@ -64,6 +85,185 @@ func (b *BtcWallet) FundPsbt(packet *psbt.Packet, minConfs int32,
+		// it's a BIP49/84 key for normal on-chain funds or a key of the
+		// custom purpose 1017 key scope.
+		derivationInfo := in.Bip32Derivation[0]
+		privKey, err := b.deriveKeyByBIP32Path(derivationInfo.Bip32Path)
```
We may not have the private key for this input, will this throw an error?

guggero replied:
Ah, you're right, we said we would only sign what we know in the godoc. Okay, added a logger to log a warning but then continue with the other outputs.

yyforyongyu commented on lnwallet/btcwallet/psbt.go:
```diff
@@ -64,6 +85,185 @@ func (b *BtcWallet) FundPsbt(packet *psbt.Packet, minConfs int32,
+			in.SighashType, privKey,
+		)
+		if err != nil {
+			return fmt.Errorf("error signing input %d: %v", idx,
```
Looks like we are signing all the inputs that have no witness and return error when signing fails. However from the docs I think we want to continue the process and sign whatever we can sign without returning an error.

yyforyongyu replied:
Or am I misunderstanding? That this method requires the ability to sign all inputs inside the psbt?

guggero replied:
If we have the key that is required to sign (checked above through the BIP32 derivation info) then we actually want to error out here if we cannot sign. Added a comment to clarify.

yyforyongyu commented on lnwallet/rpcwallet/rpcwallet.go:
```diff
@@ -651,109 +525,163 @@ func (r *RPCKeyRing) SignOutputRaw(tx *wire.MsgTx,
+	switch {
+	// No error, we do have the full UTXO and derivation info available.
+	case err == nil:
+		// Great,
```
😂

lightninglabs-deploy commented:
@guggero, remember to re-request review from reviewers for your latest update

guggero commented on lnwallet/btcwallet/signer.go:
```diff
@@ -73,6 +76,110 @@ func deriveFromKeyLoc(scopedMgr *waddrmgr.ScopedKeyManager,
 }
 
+// deriveKeyByBIP32Path derives a key described by a BIP32 path. The path is
+// expected to have 2^31 added to a hardened key in the path.
```
Not exactly. Updated the comment to be more precise.

guggero commented:
Thank you for the reviews, @carlaKC and @yyforyongyu. I addressed all comments and added a health check. Will need to do some manual re-tests to confirm they are working (will do so ASAP).

@Roasbeef in the meantime, do you mind taking a look as well?

guggero commented:
Added some health checks and critical logger in case the remote signing doesn't work.

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@carlakc: review reminder

carlaKC commented on lncfg/remotesigner.go:
```diff
@@ -1,9 +1,18 @@
+	RPCHost      string        `long:"rpchost" description:"The remote signer's RPC host:port"`
+	MacaroonPath string        `long:"macaroonpath" description:"The macaroon to use for authenticating with the remote signer"`
+	TLSCertPath  string        `long:"tlscertpath" description:"The TLS certificate to use for establishing the remote signer's identity"`
+	Timeout      time.Duration `long:"timeout" description:"The timeout for connecting to and signing requests with the remote signer. Valid time units are {s, m, h}."`
```
Validate >0?

carlaKC reviewed (approved):
New health check looks good!

Roasbeef commented on lnwallet/btcwallet/psbt.go:
```diff
@@ -64,6 +85,192 @@ func (b *BtcWallet) FundPsbt(packet *psbt.Packet, minConfs int32,
+			derivationInfo.PubKey,
+			privKey.PubKey().SerializeCompressed(),
+		)
+		if !pubKeysEqual {
```
Nice consistency check!

Roasbeef commented on config_builder.go:
```diff
@@ -687,7 +687,7 @@ func (d *RPCSignerWalletImpl) BuildChainControl(
 	rpcKeyRing, err := rpcwallet.NewRPCKeyRing(
 		baseKeyRing, walletController,
-		d.DefaultWalletImpl.cfg.RemoteSigner,
+		d.DefaultWalletImpl.cfg.RemoteSigner, walletConfig.CoinType,
```
Great commit message here, really clear how everything fits together now! Given we can just fully specify everything, the signer itself can really just be stateless other than the initial HD key info 😎

Roasbeef commented on lnwallet/rpcwallet/rpcwallet.go:
```diff
@@ -105,189 +106,7 @@ func NewRPCKeyRing(watchOnlyKeyRing keychain.SecretKeyRing,
-
-	// We need to make sure we've derived the same address on the remote
-	// signing machine, otherwise we don't know whether we're at the same
-	// address index (and therefore the same wallet state in general).
```
Yay deleted code!

Roasbeef reviewed (changes requested):
Solid set of changes! 

Only minor comment is adding some additional defensive logic there re hardened key derivation up the element of the path. 

Trying to think of other alternatives to pre-creating that set of accounts, but in the end, the tradeoff w/ the current approach isn't _too_ severe IMO.

lightninglabs-deploy commented:
@yyforyongyu: review reminder

guggero commented on lnwallet/btcwallet/signer.go:
```diff
@@ -73,6 +76,114 @@ func deriveFromKeyLoc(scopedMgr *waddrmgr.ScopedKeyManager,
+			expectedDerivationPathDepth, len(path))
+	}
+
+	purpose := path[0] - hdkeychain.HardenedKeyStart
```
Good idea, added an assertion and a test case for it.

guggero commented on lnwallet/btcwallet/log.go:
```diff
@@ -0,0 +1,32 @@
+)
+
+// Subsystem defines the logging code for this subsystem.
+const Subsystem = "BTWL"
```
To avoid a circular package dependency I wasn't able to add it to the existing logger in `lnwallet/log.go`. So I decided to also give it its own subsystem name to distinguish it from the `btcsuite/btcwallet` package.

lightninglabs-deploy commented:
@roasbeef: review reminder
@yyforyongyu: review reminder

yyforyongyu commented on lnwallet/rpcwallet/rpcwallet.go:
```diff
@@ -401,48 +219,146 @@ func (r *RPCKeyRing) FinalizePsbt(packet *psbt.Packet, accountName string) error
+		// We can only sign if we have UTXO information available. We
+		// can just continue here as a later step will fail with a more
+		// precise error message.
+		if in.WitnessUtxo == nil && in.NonWitnessUtxo == nil {
```
Can we have both as non-nil?

guggero replied:
Yes, by default we always set both in `lnd`. Some hardware wallets even require both to be set for a witness input (there was a CVE a while ago that made everyone patch the signing software to require the full non-witness UTXO information to be present).

yyforyongyu commented on lnwallet/rpcwallet/rpcwallet.go:
```diff
@@ -651,109 +525,162 @@ func (r *RPCKeyRing) SignOutputRaw(tx *wire.MsgTx,
+	// Make sure we actually know about the input. We either have been
+	// watching the UTXO on-chain or we have been given all the required
+	// info in the sign descriptor.
+	info, err := r.WalletController.FetchInputInfo(&txIn.PreviousOutPoint)
```
Do we need to check the err against `ErrNotMine` to make sure it's saying we don't have it?

guggero replied:
The `FetchInputInfo` is just used to also get the full previous TX for the `NonWitnessUtxo` field, which might be required by hardware wallets. Maybe you could say this is a premature optimization to make the code more generic.
But in most cases we populate what is required in the sign descriptor and don't actually need the full input info. So we don't really care what error is returned as we have a fallback in most cases anyway.

yyforyongyu reviewed (commented):
Lastest change looks good. Only got a few questions. And what's the plan on the itest part? Have to say this is quite an addition to psbt flow, great work!

guggero commented on lnwallet/btcwallet/psbt.go:
```diff
@@ -64,6 +85,192 @@ func (b *BtcWallet) FundPsbt(packet *psbt.Packet, minConfs int32,
+// perform any other tasks (such as coin selection, UTXO locking or
+// input/output/fee value validation, PSBT finalization). Any input that is
+// incomplete will be skipped.
+func (b *BtcWallet) SignPsbt(packet *psbt.Packet) error {
```
Both `FundPsbt` and `FinalizePsbt` only act on "normal" (BIP49/84) wallet funds and they require the wallet to know the UTXO. That's why we implemented them in `btcwallet` as they need access to the internal UTXO database.
The `SignPsbt` step doesn't require any UTXO knowledge and also supports signing with custom `lnd` specific keys (those in the `m/1017'` scope). So it made sense to keep that code in `lnd` itself.

yyforyongyu commented on lnwallet/rpcwallet/rpcwallet.go:
```diff
@@ -401,48 +219,146 @@ func (r *RPCKeyRing) FinalizePsbt(packet *psbt.Packet, accountName string) error
+		}
+
+		// Fall back to witness UTXO only for older wallets.
+		if in.WitnessUtxo != nil {
```
I guess when `in.WitnessUtxo` and `in.NonWitnessUtxo` are both non-nil, we will always end up here and overwrite the `signOutput`?

guggero replied:
Yes, that's correct. `in.WitnessUtxo` is the "most correct" one so we take that if available and overwrite any other value we might have.

guggero commented:
I added an additional commit (e79d82c1) that allows an existing wallet to be converted into a watch-only wallet to support migrating existing nodes to a remote signing setup.

carlaKC commented:
> I added an additional commit (e79d82c) that allows an existing wallet to be converted into a watch-only wallet to support migrating existing nodes to a remote signing setup.

Tested this out locally, goods as advertised 🔥 🏅

guggero commented on lnwallet/btcwallet/signer_test.go:
```diff
@@ -0,0 +1,241 @@
+	// which is a special case for the BIP49/84 addresses in btcwallet).
+	firstAddress = "bcrt1qgdlgjc5ede7fjv350wcjqat80m0zsmfaswsj9p"
+
+	testCases = []struct {
```
I created them myself by deriving the keys with another software (an online, JS based tool I wrote a few years ago). So they might not be the most creative test vectors :sweat_smile:

guggero commented on lnwallet/btcwallet/psbt.go:
```diff
@@ -64,6 +85,192 @@ func (b *BtcWallet) FundPsbt(packet *psbt.Packet, minConfs int32,
+
+// serializeTxWitness return the wire witness stack into raw bytes.
+func serializeTxWitness(txWitness *wire.TxWitness) ([]byte, error) {
+	if txWitness == nil {
```
We currently only use this in the tests, so I moved it there and removed the nil check and passed by value. It can still be nil though since it's just a type alias for `[][]byte`. But that's handled correctly in the `psbt.WriteTxWitness()` function.

yyforyongyu commented on lnwallet/btcwallet/btcwallet.go:
```diff
@@ -287,15 +287,38 @@ func (b *BtcWallet) InternalWallet() *base.Wallet {
+
+	// If the wallet is watch-only, but we don't expect it to be, then we
+	// are in an unexpected state and cannot continue.
+	if walletIsWatchOnly && !b.cfg.WatchOnly {
```
Not sure that I understand this. According to the database the wallet is watch-only, yet the config is not? How could that be? Meanwhile shouldn't these two values always be equal?

guggero replied:
The config here is coming from `--remotesigner.enable`. So if your wallet was created as watch-only but you forgot to turn on `--remotesigner.enable`, then `lnd` will inevitably crash later on because it is missing private keys and can't sign anything. So rather than crashing with a signing error, we return a nice error here. So this is purely to prevent a wrong configuration by the user.

yyforyongyu commented on lnwallet/btcwallet/btcwallet.go:
```diff
@@ -343,6 +366,23 @@ func (b *BtcWallet) Start() error {
+				"purging all private key material")
+
+			ns := tx.ReadWriteBucket(waddrmgrNamespaceKey)
+			err = b.wallet.Manager.ConvertToWatchingOnly(ns)
```
ok so previously we don't mark the wallet as watch-only in db?

guggero replied:
Previously a new wallet would be created by `lncli createwatchonly` which set this flag at wallet creation time. If we migrate a wallet that previously had private keys, then this call will set the `watch-only` boolean in the DB, yes.

yyforyongyu replied:
Got it. With this commit I guess we can now turn a non-watch-only node into a watch-only node.

yyforyongyu reviewed (commented):
Almost looks good to me, have two questions.

- what's the purpose of the last commit? That we have previously created watch-only wallets but have private keys in them?

- is there an easy way to add itest for `SignPsbt`?

guggero commented:
> what's the purpose of the last commit? That we have previously created watch-only wallets but have private keys in them?

Before this commit we assumed that you would start a remote signing setup by creating a completely _new_ node. The watch-only node would be initialized with `lncli createwatchonly` which creates a watch-only wallet from the beginning.
This new commit makes it possible to split an existing, fully signing standalone node into a remote signing setup by migrating the existing wallet to a watch-only one.

> is there an easy way to add itest for `SignPsbt`?

That RPC is indirectly tested during the remote signing tests. But you're right, having an explicit test wouldn't hurt. Going to add one.

guggero commented:
I added an itest for `SignPsbt` in 8994d86a. @yyforyongyu PTAL.

yyforyongyu reviewed (approved):
LGTM🥳 The added itest is very helpful as now I just treat the itest as example usages for all the endpoints. Again very nice job!!

Roasbeef reviewed (approved):
LGTM 🐊

Great touch with the new flag to migrate the wallet to a watch only version, this'll make it _much_ easier to production systems to retroactively adopt the remote signer architecture.

joostjager commented:
It's a nice step forward to be able to isolate the key material in a minimal lnd instance, but what's your thought on adding policy controls to prevent a hacker from using the signing instance in an unlimited way? If there are no limits, it's almost as if the hacker owns the keys themselves I'd think.

The current remote signing API is low-level and I am not sure if it is possible to add useful limits.

The lightning signer project for example works on higher level: https://gitlab.com/lightning-signer/rust-lightning-signer/-/blob/master/lightning-signer-server/src/server/remotesigner.proto

Roasbeef commented:
>  but what's your thought on adding policy controls to prevent a hacker from using the signing instance in an unlimited way? If there are no limits, it's almost as if the hacker owns the keys themselves I'd think.

This'll take place at a higher level. With the current architecture, it just so happens that the [`rpcwallet`](https://github.com/lightningnetwork/lnd/blob/master/lnwallet/rpcwallet/rpcwallet.go) is checked _into_ the `lnd` repo. It's possible for someone to either modify that package, or insert _another_ gRPC service with the same interface in front of it (applying proxy checks then denying or forwarding the request along). This service can implement the rate limiting or policy controls, unbeknownst to `lnd`. Verification of certain actions will require minimally adding additional fields to the `SignDescriptor` proto we send out. 

The purpose of the lower level API was always to enable additional verification at a higher level. Check out this old lnd ML post on the roadmap we've been following here: https://groups.google.com/a/lightning.engineering/g/lnd/c/3rF65UxvDtM/m/ktDLAGtlAQAJ.

joostjager commented:
I understand that you can have another grpc service with the same interface that acts as the remote signer. What I am wondering is what needs to be added to that interface in order to implement more advanced signing policies. For example an hourly limit on the total amount of all htlcs offered on a particular channel. I don't see this mentioned in the ML post.

If I understand it correctly, you are suggesting to extend the `SignDescriptor` to include metadata that is required for enforcing such policies? 

Also from the lightning-signer project, this is an interesting list of potential policies: https://gitlab.com/lightning-signer/docs/-/blob/master/policy-controls.md

Have you considered implementing the lightning-signer interface so that that remote signing service can be used as is?

Roasbeef commented:
> If I understand it correctly, you are suggesting to extend the SignDescriptor to include metadata that is required for enforcing such policies?

Yeah exactly. The policy logic needs to have essentially a mirrored view of the commitment state, this information can be passed over in the `SignDescriptor` messages, with it being given the initial channel state (connected to an existing node), or it identifying which signing operations are funding related (can be identified via the key family used, the PSBT has all this information. 

You may want to chat w/ the Voltage team, last I heard they're working on some draft code to start to implement policies in the way I've outlined above. The upside with the way `lnd` is laid out, is that we can re-use most of the code in the `lnwallet` and `hltcswitch` packages, so we won't need to re-write all the logic from scratch. Ultimately, you'll end up with a mini-lnd instance that spots checks everything that comes by, using all the same routines as base lnd.

> Have you considered implementing the (outgoing) lightning-signer interface in lnd so that that remote signing service can be used as is?

Last I checked their repo, they have an adapter for `lnd`, but it hasn't been touched in a few years, and they ended up renaming a bunch of things unnecessarily, which caused breaking changes. AFAICT now, the team has sort of unofficially been rolled into Spiral based on where they spend their time today. 

One difference with the lower level vs the higher level approach is that the higher level approach makes more assumptions w.r.t how the protocol looks _today_. For example see this call: `SignJusticeSweep`. Post anchors, it's possible that the signer needs to re-spend the justice transaction several times as the "attacker" has more freedom w.r.t going to the second level, doing other sub-aggregations, etc, etc. The lower level approach lets us keep that logic in `lnd` ([handling all the various breach spending sub-cases](https://github.com/lightningnetwork/lnd/blob/master/contractcourt/breacharbiter.go#L572)), while keeping the exposed API (sign this w/ this info) more agnostic. AFAICT, the interface doesn't handle these cases as LDK+CL haven't yet implemented anchors. Also consider that interface will also need to nearly entirely redone when we move over to taproot based channels.

joostjager commented:
Still, I wonder if the high-level approach with many different calls is genuinely different from lnd's low-level API with metadata in sign descriptors. Won't you end up packing the same data that lightning-signer has in function arguments into the metadata?

It would be nice if at least some code is usable across implementations, but it seems that, unfortunately, the work that the lightning-signer team has done needs to be replicated in full based on `lnwallet` and `htlcswitch`?

Roasbeef commented:
> It's feasible that someone can just make an API bridge that takes our lower level messages, and annotates them properly for the lightning-signer project. 

With this PR, we started to define some custom PSBT tags, like the stuff needed to do the private key operations to send for revocation keys. Continuing down this line, an alternative path could be to use more custom tags to provide the extra annotation information. This can be passed via functional arguments to the main signing interface, or by having the caller set the additional new fields in the `SignDescriptor`. This could be an interesting path, as then we can standardize things around PSBT fields w/ additional signing context, vs a slightly more implementation specific gRPC interface.

devrandom commented:
Just a quick note after reading the above.

We've been heads-down working on a beta release and also upstreaming changes to C-Lightning for integration.  C-L is pretty close to LDK for signer architecture and API, so it was the low-hanging fruit.  But we're definitely interested in moving the LND integration forward.

I do have some concerns about the metadata approach.  For one, I feel that approach is less "type safe", there seem to be more opportunities for breakage compared to an API that can be checked at compile-time.  Another issue is that new calls are required for a validating signer.  For example, the signer needs to see that the counterparty signed the current tx and revoked the old one, otherwise the state didn't really move forward.  Providing these pieces of information requires new calls that are not actual signing operation.  Also doesn't seem to fit the PSBT paradigm.

About `SignJusticeSweep` - the method takes a tx and an input index, so the node is free to construct txs and re-sign them as needed.  I do agree that there will be some API changes as the protocol is revised, e.g. with taproot.  But changes will be required in the metadata approach too - different pieces of information made available to the signer.  It just seems to be irreducible complexity.

Finally, I want to mention that we are renaming the project to Validating Lightning Signer (VLS).

joostjager commented:
Created new issue where we can continue the discussion: https://github.com/lightningnetwork/lnd/issues/6243

Roasbeef commented:
@devrandom replying here, but we can move to the issue afterwards

> I do have some concerns about the metadata approach. For one, I feel that approach is less "type safe", there seem to be more opportunities for breakage compared to an API that can be checked at compile-time

It would be no less type safe, Go is a typed language, as a protos. VLS uses protos already, and this would basically just be proto translation. 

> Another issue is that new calls are required for a validating signer. For example, the signer needs to see that the counterparty signed the current tx and revoked the old one, otherwise the state didn't really move forward. Providing these pieces of information requires new calls that are not actual signing operation.

That's correct, from an implementation perspective, this could take the form of db action hooks, or ofc we can just shove everything into a fake signing operation ;). 

If you look at in from another perspective, with the way the protocol works, we can't actually sign again _until_ the remote party revokes their commitment (we don't have the revocation point needed to make a new state otherwise). As a result, the existence of that next revocation point (which can be validated) may be sufficient to validate that a state transition actually happened. 

> But changes will be required in the metadata approach too - different pieces of information made available to the signer. It just seems to be irreducible complexity.

Yep that's unavoidable, but in the `lnd` case, the slimmer signer and the full node may just be using the exact same set of libraries to construct+validate. I don't see how it's any more complex. Ultimately you need to send the same amount of information in the abstract, the question here is how that abstraction is exposed to the implementation. [Calls like this](https://gitlab.com/lightning-signer/validating-lightning-signer/-/blob/master/lightning-signer-server/src/server/remotesigner.proto#L432) already map pretty cleanly as is.
